test/integration/client/exec_test.go

```diff
@@ -0,0 +1,329 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/util/cert"
+
+	kubeapiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+// This file tests the client-go credential plugin feature.
+
+func TestExecPlugin(t *testing.T) {
+	// These constants are used to communicate behavior to the testdata/exec-plugin.sh test fixture.
+	const (
+		outputEnvVar = "EXEC_PLUGIN_OUTPUT"
+	)
+
+	const (
+		clientAuthorizedToken   = "authorized-token"
+		clientUnauthorizedToken = "unauthorized-token"
+	)
+
+	certDir, err := ioutil.TempDir("", "kubernetes-client-exec-test-cert-dir-*")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tokenFileName := writeTokenFile(t, clientAuthorizedToken)
+	clientCAFileName, clientSigningCert, clientSigningKey := writeCACertFiles(t, certDir)
+	clientCertFileName, clientKeyFileName := writeCerts(t, clientSigningCert, clientSigningKey, certDir, 30*time.Second)
+	result := kubeapiservertesting.StartTestServerOrDie(
+		t,
+		nil,
+		[]string{
+			"--token-auth-file", tokenFileName,
+			"--client-ca-file=" + clientCAFileName,
+		},
+		framework.SharedEtcd(),
+	)
+	defer result.TearDownFn()
+
+	unauthorizedCert, unauthorizedKey, err := cert.GenerateSelfSignedCertKey("some-host", nil, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name                  string
+		clientConfigFunc      func(*rest.Config)
+		wantClientErrorPrefix string
+	}{
+		{
+			name: "unauthorized token",
+			clientConfigFunc: func(c *rest.Config) {
+				c.ExecProvider.Env = append(c.ExecProvider.Env, clientcmdapi.ExecEnvVar{
+					Name: outputEnvVar,
+					Value: `{
+										"kind": "ExecCredential",
+										"apiVersion": "client.authentication.k8s.io/v1beta1",
+										"status": {
+											"token": "unauthorized"
+										}
+									}`,
+				})
+			},
+			wantClientErrorPrefix: "Unauthorized",
+		},
+		{
+			name: "unauthorized certificate",
+			clientConfigFunc: func(c *rest.Config) {
+				c.ExecProvider.Env = append(c.ExecProvider.Env, clientcmdapi.ExecEnvVar{
```

Comment:
nit: using `append` here seems wrong - we should just set the slice to what we expect it to be.

Reply:
12dc50eff84ced8632cd0bd42a0f26a52007da45